### DISCUSSION TITLE
Strange require() not supported, @auth/express with Typescript error

### DISCUSSION THREAD

**Discussion by lucasrothman** [nextauthjs/next-auth#11417] [Help]
Throwing an exception when trying to use auth/express with Typescript. 

````
\api\node_modules\ts-node\dist\index.js:851
            return old(m, filename);
                   ^
Error [ERR_REQUIRE_ESM]: require() of ES Module \api\node_modules\@auth\express\index.js from \api\server.ts not supported.
Instead change the require of index.js in \api\server.ts to a dynamic import() which is available in all CommonJS modules.
    at require.extensions.<computed> [as .js] (\api\node_modules\ts-node\dist\index.js:851:20)
    at Object.<anonymous> (\api\server.ts:9:19)
    at m._compile (\api\node_modules\ts-node\dist\index.js:857:29)
    at require.extensions.<computed> [as .ts] (\api\node_modules\ts-node\dist\index.js:859:16)
    at phase4 (\api\node_modules\ts-node\dist\bin.js:466:20)
    at bootstrap (\api\node_modules\ts-node\dist\bin.js:54:12)
    at main (\api\node_modules\ts-node\dist\bin.js:33:12)
    at Object.<anonymous> (\api\node_modules\ts-node\dist\bin.js:579:5) {
  code: 'ERR_REQUIRE_ESM'
````

`server.ts`:
````
import express from "express";
import cors from "cors";
import { ExpressAuth } from "@auth/express";
import GitHub from "@auth/express/providers/github";

import api from "./routes/api";

const app = express();
const port = process.env.PORT || 3001;

app.use(cors());
app.use(express.json());
app.use("/auth/*", ExpressAuth({ providers: [GitHub] }));
````

`ts-config.json`:

````
{
    "compilerOptions": {
      "target": "es6",
      "module": "commonjs",
      "outDir": "./dist",
      "rootDir": "./",
      "strict": true,
      "esModuleInterop": true,
      "skipLibCheck": true,
      "forceConsistentCasingInFileNames": true
    }
  }
````

Run with `ts-node server.ts`

Potentially related to https://stackoverflow.com/questions/70800567/return-oldm-filename-error-err-require-esm-require-of-es-module but I can't seem to fix. Any thoughts?

**Comment by hillac** [comment]
Does your package.json have "type": "module"?. You might also need NodeNext instead of commonjs in tsconfig for module. Check the [example app](https://github.com/nextauthjs/express-auth-example/blob/main/src/app.ts)

You will also need to add .js file extensions to imports of your own files: `import api from "./routes/api.js";` (even though theyre .ts). It's required in ESM.

Lastly, to run the dev server, use [tsx](https://www.npmjs.com/package/tsx) instead of ts-node. It's probably the easiest way to run typescript in node and faster than ts-node (mainly because it doesn't type check). Just using tsx will probably fix all your issues anyway without having to change imports and configs. It's still probably worth fixing that stuff so you can do a normal tsc compile and run with node in production, because you shouldn't use tsx in prod. Otherwise, if you dont want to change your import and configs, you might be able to compile with [pkgroll](https://tsx.is/compilation), but I've never tried. For a new project definitely just use ESM.

If you want type checking when running the dev server, you have to use the ts-node esm loader:
`node --loader ts-node/esm server.ts`
Personally, I use [ts-node with swc](https://typestrong.org/ts-node/docs/swc/) as it's _slightly_ faster on startup comapred to tsx, which is actually noticeable imo when you're using watch mode. However, like tsx, it has no type checking, so i just run `tsc --noEmit` before committing.
Add: `"ts-node": { "swc": true}` to your tsconfig, and `npm i -D @swc/core`
Then run with `node --watch --loader ts-node/esm server.ts`

**Reply by hillac** [reply]
Probably worth putting all this info in the docs
